src/Host/Broker/Impl/Pipes/MessagePipe.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.Extensions.Logging;
+using Microsoft.R.Host.Client;
+
+namespace Microsoft.R.Host.Broker.Pipes {
+    public class MessagePipe {
+        private readonly ILogger _logger;
+        private readonly BufferBlock<byte[]> _hostMessages = new BufferBlock<byte[]>();
```

Comment:
Side note: this is likely to be the bottleneck, since all messages between host and VS flow through here. If there's some implementation of an async concurrent pipe or pipe-like structure that has better perf than BufferBlock, or if we can hand-roll such a thing, it's probably worth investigating.